src/litdata/utilities/base.py

```diff
@@ -0,0 +1,139 @@
+# Copyright The Lightning AI team.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Iterator, List, Optional
+
+from torch.utils.data import IterableDataset
+
+from litdata.streaming.dataset import StreamingDataset
+
+__NUM_CYCLES_KEY__ = "__NUM_CYCLES__"
+__NUM_SAMPLES_YIELDED_KEY__ = "__NUM_SAMPLES_YIELDED__"
+__SAMPLES_KEY__ = "__SAMPLES__"
+
+
+class _BaseStreamingDatasetWrapper(IterableDataset, ABC):
+    # Base class for datasets that wrap multiple streaming datasets
+    # This includes CombinedStreamingDataset and ParallelStreamingDataset
+
+    _datasets: List[StreamingDataset]
+    _iterator: "Optional[_BaseDatasetWrapperIterator]"
+    _current_epoch: int
+    batch_size: int
+    num_workers: int
+    _force_override_state_dict: bool
+    _use_streaming_dataloader: bool
+    _num_samples_yielded: Optional[List[int]]
+
+    def set_shuffle(self, shuffle: bool) -> None:
+        """Set the current shuffle to the datasets."""
+        for dataset in self._datasets:
+            dataset.set_shuffle(shuffle)
+
+    def set_batch_size(self, batch_size: int) -> None:
+        """Set the current batch size to the datasets."""
+        self.batch_size = batch_size
+        for dataset in self._datasets:
+            dataset.set_batch_size(batch_size)
+
+    def set_num_workers(self, num_workers: int) -> None:
+        """Set the current number of workers to the datasets."""
+        self.num_workers = num_workers
+        for dataset in self._datasets:
+            dataset.set_num_workers(num_workers)
+
+    def set_drop_last(self, drop_last: bool) -> None:
+        """Set the current drop_last to the datasets."""
+        for dataset in self._datasets:
+            dataset.set_drop_last(drop_last)
+
+    def reset_state_dict(self) -> None:
+        """Reset the state of the dataset."""
+        for dataset in self._datasets:
+            dataset.reset_state_dict()
+
+    def _check_datasets(self, datasets: List[StreamingDataset]) -> None:
+        if any(not isinstance(d, StreamingDataset) for d in datasets):
```

Comment:
```suggestion
        if any(not isinstance(d, (StreamingDataset, CombinedStreamingDataset)) for d in datasets):
```

I think ParallelSD can also include CombinedSD. Any potential pitfalls with that?